posts/2019-10-28-nll-hard-errors.md

```diff
@@ -0,0 +1,107 @@
+---
+layout: post
+title: "Completing the transition to the new borrow checker"
+author: Niko Matsakis
+---
+
+For most of 2018, we've been issuing warnings about various bugs in the
+borrow checker that we plan to fix -- about two months ago, in the
+current Rust nightly, those warnings became **hard errors**. In about
+two weeks, when the nightly branches to become beta, those hard errors
+will be in the beta build, and they will eventually hit stable on
+December 19th, as part of Rust 1.40.0. **If you're testing with
+Nightly, you should be all set -- but otherwise, you may want to go
+and check to make sure your code still builds. If not, we have advice
+for fixing common problems below.**
+
+### Background: the non-lexical lifetime transition
+
+When we [released Rust 2018 in Rust 1.31][2018], it included a new
+version of the borrow checker, one that implemented ["non-lexical
+lifetimes"][nll]. This new borrow checker did a much more precise
+analysis than the original, allowing us to eliminate a lot of
+unnecessary errors and make Rust easier to use. I think most everyone
+who was using Rust 2015 can attest that this shift was a big
+improvement.
+
+### The new borrow checker also fixed a lot of bugs
+
+What is perhaps less well understood is that the new borrow checker
+implementation *also* fixed a lot of bugs. In other words, the new
+borrow checker did not just accept more programs -- **it also rejected
+some programs that never should have been accepted in the first
+place!**
```

Comment:
Maybe mention that these were soundness bugs ("memory unsafety bugs" might work better).